### PR TITLE
#2367 dapr change the app port depending on the app protocol

### PR DESCRIPTION
**Pull request by paule96**
if you dapr connected apps uses anything else then `http` they will currently not correctly connecting to the sidecare. (more details to that issue in #2367 )

To fix this now, a use can set the app protocol in the sidecare options and this will then take the correct endpoint information.

Open questions:

- [x] Can someone test the playground project for dapr?
  - for me on Linux it seems to be broken in

 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/dotnet/aspire/pull/3184)

### CODE REVIEW

**Comment by paule96** [comment]
okay, I know now why the playground project is not working. There seems to be a bug unrelated to this PR that it is required to have dapr `1.13.0` as a reference for aspire >= preview 4.
The problem is there is only dapr `1.12.0` available on the feeds configured in the NuGet.config.

**Comment by paule96** [comment]
Okay after I integrated locally the nuget.org feed I was able to test my change. For me it looks like it is working.

File: src/Aspire.Hosting.Dapr/DaprDistributedApplicationLifecycleHook.cs [hunk 1]
```diff
@@ -146,13 +146,13 @@ public async Task BeforeStartAsync(DistributedApplicationModel appModel, Cancell
...
+                        
                         context.EnvironmentVariables.TryAdd("DAPR_GRPC_ENDPOINT", grpc);
                         context.EnvironmentVariables.TryAdd("DAPR_HTTP_ENDPOINT", http);
                     }));
 
-            daprCli.Annotations.Add(new EndpointAnnotation(ProtocolType.Tcp, name: "grpc", port: sidecarOptions?.DaprGrpcPort));
-            daprCli.Annotations.Add(new EndpointAnnotation(ProtocolType.Tcp, name: "http", port: sidecarOptions?.DaprHttpPort));
+            daprCli.Annotations.Add(new EndpointAnnotation(ProtocolType.Tcp, name: "grpc", port: sidecarOptions?.DaprGrpcPort, uriScheme: "http"));
+            daprCli.Annotations.Add(new EndpointAnnotation(ProtocolType.Tcp, name: "http", port: sidecarOptions?.DaprHttpPort, uriScheme: "http"));
             daprCli.Annotations.Add(new EndpointAnnotation(ProtocolType.Tcp, name: "metrics", port: sidecarOptions?.MetricsPort));
```
Comment (paule96) [inline src/Aspire.Hosting.Dapr/DaprDistributedApplicationLifecycleHook.cs:156]: @philliphoff do you know what the uri schema for metrics and profiling should be?

Comment (philliphoff) [reply src/Aspire.Hosting.Dapr/DaprDistributedApplicationLifecycleHook.cs:156]: Both endpoints are HTTP-based, with the metrics endpoint being a basic HTTP endpoint and, I believe, the profiling endpoint being gRPC.

Comment (paule96) [reply src/Aspire.Hosting.Dapr/DaprDistributedApplicationLifecycleHook.cs]: found that in the docs for the profiling option:

| Name | Environment Variable |  Default | Description |
| ------ | ------------------------  | ------- | ------------- |
| --enable-profiling |   | false | Enable “pprof” profiling via an HTTP endpoint |


So it's HTTP too. Changed it accordingly.

**Comment by paule96** [comment]
@davidfowl and @karolz-ms,

because you both were involved in #3305 can I ask you to review this PR too?

File: src/Aspire.Hosting.Dapr/DaprDistributedApplicationLifecycleHook.cs [hunk 2]
```diff
@@ -146,17 +146,16 @@ public async Task BeforeStartAsync(DistributedApplicationModel appModel, Cancell
 
                         var http = daprCli.GetEndpoint("http");
                         var grpc = daprCli.GetEndpoint("grpc");
-
+
```
Comment (davidfowl) [inline src/Aspire.Hosting.Dapr/DaprDistributedApplicationLifecycleHook.cs:149]:
```suggestion
```

File: src/Aspire.Hosting.Dapr/DaprDistributedApplicationLifecycleHook.cs [hunk 3]
```diff
@@ -146,17 +146,16 @@ public async Task BeforeStartAsync(DistributedApplicationModel appModel, Cancell
 
                         var http = daprCli.GetEndpoint("http");
                         var grpc = daprCli.GetEndpoint("grpc");
-
+                        
                         context.EnvironmentVariables.TryAdd("DAPR_GRPC_ENDPOINT", grpc);
                         context.EnvironmentVariables.TryAdd("DAPR_HTTP_ENDPOINT", http);
                     }));
-
```
Comment (davidfowl) [inline src/Aspire.Hosting.Dapr/DaprDistributedApplicationLifecycleHook.cs:153]: Why did you remove the space?

File: src/Aspire.Hosting.Dapr/DaprDistributedApplicationLifecycleHook.cs [hunk 4]
```diff
@@ -174,7 +173,8 @@ public async Task BeforeStartAsync(DistributedApplicationModel appModel, Cancell
                         EndpointReference? httpEndPoint = null;
                         if (resource is IResourceWithEndpoints resourceWithEndpoints)
                         {
-                            httpEndPoint = resourceWithEndpoints.GetEndpoint("http");
+                            var endpointName = sidecarOptions?.AppProtocol ?? "http";
```
Comment (davidfowl) [inline src/Aspire.Hosting.Dapr/DaprDistributedApplicationLifecycleHook.cs:176]: This coupling between the dapr protocol and the endpoint name of the app is interesting. Maybe we should have 2 separate options and a default fallback where we assume the app protocol and endpoint name are the same.

File: src/Aspire.Hosting.Dapr/DaprDistributedApplicationLifecycleHook.cs [hunk 5]
```diff
@@ -146,17 +146,16 @@ public async Task BeforeStartAsync(DistributedApplicationModel appModel, Cancell
...
                         context.EnvironmentVariables.TryAdd("DAPR_HTTP_ENDPOINT", http);
                     }));
-
             daprCli.Annotations.Add(new EndpointAnnotation(ProtocolType.Tcp, uriScheme: "http", name: "grpc", port: sidecarOptions?.DaprGrpcPort));
             daprCli.Annotations.Add(new EndpointAnnotation(ProtocolType.Tcp, uriScheme: "http", name: "http", port: sidecarOptions?.DaprHttpPort));
             daprCli.Annotations.Add(new EndpointAnnotation(ProtocolType.Tcp, uriScheme: "http", name: "metrics", port: sidecarOptions?.MetricsPort));
             if (sidecarOptions?.EnableProfiling == true)
             {
-                daprCli.Annotations.Add(new EndpointAnnotation(ProtocolType.Tcp, name: "profile", port: sidecarOptions?.ProfilePort));
+                daprCli.Annotations.Add(new EndpointAnnotation(ProtocolType.Tcp, name: "profile", port: sidecarOptions?.ProfilePort, uriScheme: "http"));
```
Comment (davidfowl) [inline src/Aspire.Hosting.Dapr/DaprDistributedApplicationLifecycleHook.cs:158]: Good catch!

File: playground/dapr/AppHost/Program.cs [hunk 6]
```diff
@@ -1,15 +1,21 @@
+using Aspire.Hosting.Dapr;
```
Comment (davidfowl) [inline playground/dapr/AppHost/Program.cs:1]: We don't want to change the sample, can you write a test instead? We now have dapr tests.

Comment (paule96) [reply playground/dapr/AppHost/Program.cs:1]: cool was not sure. When I started with that PR there were no tests for dapr. :)

File: src/Aspire.Hosting.Dapr/DaprDistributedApplicationLifecycleHook.cs (hunk 4 above)
Comment (paule96) [reply src/Aspire.Hosting.Dapr/DaprDistributedApplicationLifecycleHook.cs:175]: @davidfowl you mean something like that:

```csharp
builder.AddProject<Projects.DaprServiceA>("servicea", "https")
       .WithDaprSidecar(new DaprSidecarOptions{
              AppEndpoint = "MyEndpoint",
              AppProtocol = "HTTP"
       })
```

And there are the following cases:

- `AppEndpoint` and `AppProtocol ` is `null`
  - use `http` for both
- `AppEndpoint` is **not** `null` and `AppProtocol` is `null`
  - use as the `AppProtocol` the protocol that is defined by the `AppEndpoint` URL. 
  - if the extraction failes, fall back to `http`
- `AppEndpoint` is `null` and `AppProtocol` is **not** `null`
  - Get the app endpoint with the `AppProtocol` name and set it as `AppEndpoint`
  - if the extraction failes, fall back to `http` endpoint
    - if that fails exception??
- `AppEndpoint` is not `null` and `AppProtocol` is not `null`
  - use both as they are defined

File: src/Aspire.Hosting.Dapr/DaprDistributedApplicationLifecycleHook.cs (hunk 5 above)
Comment (paule96) [reply src/Aspire.Hosting.Dapr/DaprDistributedApplicationLifecycleHook.cs:157]: yes we both fixed the same thing. :)

File: src/Aspire.Hosting.Dapr/DaprDistributedApplicationLifecycleHook.cs (hunk 3 above)
Comment (paule96) [reply src/Aspire.Hosting.Dapr/DaprDistributedApplicationLifecycleHook.cs]: fixed

File: src/Aspire.Hosting.Dapr/DaprDistributedApplicationLifecycleHook.cs (hunk 4 above)
Comment (paule96) [reply src/Aspire.Hosting.Dapr/DaprDistributedApplicationLifecycleHook.cs]: resolved

**Comment by paule96** [comment]
@davidfowl please review again :) I changed the logic and also added tests

File: src/Aspire.Hosting.Dapr/DaprDistributedApplicationLifecycleHook.cs [hunk 7]
```diff
@@ -170,16 +171,11 @@ public async Task BeforeStartAsync(DistributedApplicationModel appModel, Cancell
...
-
-                        EndpointReference? httpEndPoint = null;
-                        if (resource is IResourceWithEndpoints resourceWithEndpoints)
-                        {
-                            var endpointName = sidecarOptions?.AppProtocol ?? "http";
-                            httpEndPoint = resourceWithEndpoints.GetEndpoint(endpointName.ToLower());
-
-                            if (httpEndPoint.IsAllocated && sidecarOptions?.AppPort is null)
+                        var endPoint = GetEndpointReference(sidecarOptions, resource);
+                        if(endPoint is not null){
```
Comment (davidfowl) [inline src/Aspire.Hosting.Dapr/DaprDistributedApplicationLifecycleHook.cs:175]: nit: Fix the formatting.

File: src/Aspire.Hosting.Dapr/DaprDistributedApplicationLifecycleHook.cs [hunk 8]
```diff
@@ -256,6 +255,33 @@ public async Task BeforeStartAsync(DistributedApplicationModel appModel, Cancell
         appModel.Resources.AddRange(sideCars);
     }
 
+    static (EndpointReference appEndpoint, string protocol)? GetEndpointReference(DaprSidecarOptions? sidecarOptions, IResource resource)
+    {
+        if (resource is IResourceWithEndpoints resourceWithEndpoints)
```
[the remark this answers is not in the document]
Comment (paule96) [reply src/Aspire.Hosting.Dapr/DaprDistributedApplicationLifecycleHook.cs:276]: done

File: src/Aspire.Hosting.Dapr/DaprDistributedApplicationLifecycleHook.cs (hunk 7 above)
Comment (paule96) [reply src/Aspire.Hosting.Dapr/DaprDistributedApplicationLifecycleHook.cs]: done

File: src/Aspire.Hosting.Dapr/DaprDistributedApplicationLifecycleHook.cs [hunk 9]
```diff
@@ -255,6 +257,47 @@ public async Task BeforeStartAsync(DistributedApplicationModel appModel, Cancell
         appModel.Resources.AddRange(sideCars);
     }
 
+    /// <summary>
+    /// Get's the applications endpoint and the protocl for it.
```
Comment (davidfowl) [inline src/Aspire.Hosting.Dapr/DaprDistributedApplicationLifecycleHook.cs:261]:
```suggestion
    /// Get's the applications endpoint and the protocol for it.
```

File: src/Aspire.Hosting.Dapr/DaprDistributedApplicationLifecycleHook.cs [hunk 10]
```diff
@@ -255,6 +257,47 @@ public async Task BeforeStartAsync(DistributedApplicationModel appModel, Cancell
...
+    /// - <see cref="DaprSidecarOptions.AppEndpoint"/>
+    /// The logic is like:
+    /// - if both are null try to use 'http' for both
+    /// - if <see cref="DaprSidecarOptions.AppProtocol"/> is not null try to get an endpoint with the name of the protocol
+    /// - if <see cref="DaprSidecarOptions.AppEndpoint"/> is not null try to use the schema as the protocol
+    /// - if bot are not null just use them.
+    /// </summary>
+    /// <param name="sidecarOptions">The current options that are the base of the logic.</param>
+    /// <param name="resource">The application resource the sidecar is connected to.</param>
+    /// <returns>The appendpoint that should be used and the protocol.</returns>
```
Comment (davidfowl) [inline src/Aspire.Hosting.Dapr/DaprDistributedApplicationLifecycleHook.cs:273]:
```suggestion
    // This method resolves the application's endpoint and the protocol that the dapr side car will use.
    // It depends on DaprSidecarOptions.AppProtocol and DaprSidecarOptions.AppEndpoint.
    // - If both are null default to 'http' for both.
    // - If AppProtocol is not null try to get an endpoint with the name of the protocol.
    // - if AppEndpoint is not null try to use the scheme as the protocol.
    // - if both are not null just use both options.
```

File: src/Aspire.Hosting.Dapr/DaprDistributedApplicationLifecycleHook.cs [hunk 11]
```diff
@@ -255,6 +257,39 @@ public async Task BeforeStartAsync(DistributedApplicationModel appModel, Cancell
...
+    // It depends on DaprSidecarOptions.AppProtocol and DaprSidecarOptions.AppEndpoint.
+    // - If both are null default to 'http' for both.
+    // - If AppProtocol is not null try to get an endpoint with the name of the protocol.
+    // - if AppEndpoint is not null try to use the scheme as the protocol.
+    // - if both are not null just use both options.
+    static (EndpointReference appEndpoint, string protocol)? GetEndpointReference(DaprSidecarOptions? sidecarOptions, IResource resource)
+    {
+        if (resource is IResourceWithEndpoints resourceWithEndpoints)
+        {
+            if (sidecarOptions == null || (sidecarOptions.AppProtocol == null && sidecarOptions.AppEndpoint == null))
```
Comment (davidfowl) [inline src/Aspire.Hosting.Dapr/DaprDistributedApplicationLifecycleHook.cs:270]: Can you rewrite this as a switch expression?

Comment (paule96) [reply src/Aspire.Hosting.Dapr/DaprDistributedApplicationLifecycleHook.cs:270]: @davidfowl Do you have time for a bit of guidance here.
I tried it but failed and I don't know what I do wrong:

```csharp
return (sidecarOptions, resourceWithEndpoints) switch{
                (DaprSidecarOptions? p0, IResourceWithEndpoints p1) when p0 == null || (p0.AppProtocol == null && p0.AppEndpoint == null) => (p1.GetEndpoint("http"), "http"),
                _ => throw new ArgumentException($"The combination of {nameof(DaprSidecarOptions.AppEndpoint)} and {nameof(DaprSidecarOptions.AppProtocol)}")
            }
```

That's my current code. And it will result in:

"CS0119", 'DaprSidecarOptions' is a type, which is not valid in the given context"

Comment (paule96) [reply src/Aspire.Hosting.Dapr/DaprDistributedApplicationLifecycleHook.cs:270]: @davidfowl nevermind... I just used `var` and it works...

Comment (davidfowl) [reply src/Aspire.Hosting.Dapr/DaprDistributedApplicationLifecycleHook.cs:270]: I think it would look something like this:

```C#
if (resource is not IResourceWithEndpoints resourceWithEndpoints)
{
    return null;
}

return (sidecarOptions?.AppProtocol, sidecarOptions?.AppEndpoint) switch
{
    // If both are not null, use AppEndpoint and AppProtocol
    (string appProtocol, string appEndpoint) => (resourceWithEndpoints.GetEndpoint(appEndpoint), appProtocol),

    // If both are null, default to http
    (null, null) => (resourceWithEndpoints.GetEndpoint("http"), "http"),

    // If AppProtocol is null, use AppEndpoint
    (null, string appEndpoint) => (resourceWithEndpoints.GetEndpoint(appEndpoint), appEndpoint),

    // If both are not null, use AppEndpoint and AppProtocol
    (string appProtocol, null) => (resourceWithEndpoints.GetEndpoint(appProtocol), appProtocol);
}
```

File: src/Aspire.Hosting.Dapr/DaprDistributedApplicationLifecycleHook.cs [hunk 12]
```diff
@@ -267,25 +267,13 @@ public async Task BeforeStartAsync(DistributedApplicationModel appModel, Cancell
...
-            {
-                var appEndpoint = resourceWithEndpoints.GetEndpoint(sidecarOptions.AppProtocol);
-                return (appEndpoint, sidecarOptions.AppProtocol);
-            }
-            else if (sidecarOptions.AppProtocol != null && sidecarOptions.AppEndpoint != null)
-            {
-                var appEndpoint = resourceWithEndpoints.GetEndpoint(sidecarOptions.AppEndpoint);
-                return (appEndpoint, sidecarOptions.AppProtocol);
-            }
+            return (sidecarOptions, resourceWithEndpoints) switch{
```
Comment (davidfowl) [inline src/Aspire.Hosting.Dapr/DaprDistributedApplicationLifecycleHook.cs:270]: Why not get rid of teh when expressions and just check the 2 values? That makes it much cleaner.

Comment (paule96) [reply src/Aspire.Hosting.Dapr/DaprDistributedApplicationLifecycleHook.cs:270]: how does that work? It's my first time with the new switch expression :O

Comment (RussKie) [reply src/Aspire.Hosting.Dapr/DaprDistributedApplicationLifecycleHook.cs:270]: It's the same as the old `switch` with less "boilerplate" if you like. You may find [these docs](https://learn.microsoft.com/dotnet/csharp/language-reference/operators/switch-expression) helpful.

In a nutshel, this method switches on a number of different cases (quite elaborate) where args are typles and for each case returns tuples.

Does this makes sense?

File: src/Aspire.Hosting.Dapr/DaprDistributedApplicationLifecycleHook.cs [hunk 13]
```diff
@@ -255,6 +257,27 @@ public async Task BeforeStartAsync(DistributedApplicationModel appModel, Cancell
...
+    // It depends on DaprSidecarOptions.AppProtocol and DaprSidecarOptions.AppEndpoint.
+    // - If both are null default to 'http' for both.
+    // - If AppProtocol is not null try to get an endpoint with the name of the protocol.
+    // - if AppEndpoint is not null try to use the scheme as the protocol.
+    // - if both are not null just use both options.
+    static (EndpointReference appEndpoint, string protocol)? GetEndpointReference(DaprSidecarOptions? sidecarOptions, IResource resource)
+    {
+        if (resource is IResourceWithEndpoints resourceWithEndpoints)
+        {
+            return (sidecarOptions, resourceWithEndpoints) switch{
```
Comment (RussKie) [inline src/Aspire.Hosting.Dapr/DaprDistributedApplicationLifecycleHook.cs:270]: [styling] Here and few other places the curly has to be on the new line

File: tests/Aspire.Hosting.Tests/Dapr/DaprTests.cs [hunk 14]
```diff
@@ -77,10 +77,92 @@ public async Task WithDaprSideCarAddsAnnotationAndSidecarResource()
...
+                e.UriScheme = "http";
+                e.AllocatedEndpoint = new(e, "localhost", 8000);
+            })
+            .WithEndpoint("https", e =>
+            {
+                e.Port = 8001;
+                e.UriScheme = "https";
+                e.AllocatedEndpoint = new(e, "localhost", 8001);
+            });
+        if(schema == null && endPoint == null && port == null){
```
Comment (RussKie) [inline tests/Aspire.Hosting.Tests/Dapr/DaprTests.cs:122]: [styling] spacing
```suggestion
        if (schema == null && endPoint == null && port == null) {
```

File: src/Aspire.Hosting.Dapr/DaprDistributedApplicationLifecycleHook.cs [hunk 15]
```diff
@@ -255,6 +257,27 @@ public async Task BeforeStartAsync(DistributedApplicationModel appModel, Cancell
...
+                (var p0, var p1) when p0 == null || (p0.AppProtocol == null && p0.AppEndpoint == null) => (p1.GetEndpoint("http"), "http"),
```
Comment (RussKie) [inline src/Aspire.Hosting.Dapr/DaprDistributedApplicationLifecycleHook.cs:271]: ❗  Equality operator can be overloaded which can load to hard to track bugs.
```suggestion
                (var p0, var p1) when p0 is null || (p0.AppProtocol is null && p0.AppEndpoint is null) => (p1.GetEndpoint("http"), "http"),
```

ditto other places

File: src/Aspire.Hosting.Dapr/DaprDistributedApplicationLifecycleHook.cs (hunk 12 above)
Comment (paule96) [reply src/Aspire.Hosting.Dapr/DaprDistributedApplicationLifecycleHook.cs:270]: Sadly no. I did read the documentation but have no clue how it should work with two a parameters

File: src/Aspire.Hosting.Dapr/DaprDistributedApplicationLifecycleHook.cs (hunk 15 above)
Comment (paule96) [reply src/Aspire.Hosting.Dapr/DaprDistributedApplicationLifecycleHook.cs]: done

File: tests/Aspire.Hosting.Tests/Dapr/DaprTests.cs (hunk 14 above)
Comment (paule96) [reply tests/Aspire.Hosting.Tests/Dapr/DaprTests.cs]: done

File: src/Aspire.Hosting.Dapr/DaprDistributedApplicationLifecycleHook.cs (hunk 13 above)
Comment (paule96) [reply src/Aspire.Hosting.Dapr/DaprDistributedApplicationLifecycleHook.cs]: done

File: src/Aspire.Hosting.Dapr/DaprDistributedApplicationLifecycleHook.cs (hunk 12 above)
Comment (paule96) [reply src/Aspire.Hosting.Dapr/DaprDistributedApplicationLifecycleHook.cs]: can you provide an example, maybe with the code from this pr

Comment (davidfowl) [reply src/Aspire.Hosting.Dapr/DaprDistributedApplicationLifecycleHook.cs]: My comment here described it https://github.com/dotnet/aspire/pull/3184#discussion_r1560554150

Comment (paule96) [reply src/Aspire.Hosting.Dapr/DaprDistributedApplicationLifecycleHook.cs]: sorry didn't saw this comment. Fixed it now.

**Comment by davidfowl** [comment]
This looks great

File: tests/Aspire.Hosting.Tests/Dapr/DaprTests.cs [hunk 16]
```diff
@@ -77,10 +77,96 @@ public async Task WithDaprSideCarAddsAnnotationAndSidecarResource()
...
+        var ports = new Dictionary<string, int>
+        {
+            ["http"] = 3500,
+            ["grpc"] = 50001,
+            ["metrics"] = 9090
+        };
+
+        foreach (var e in endpoints)
+        {
+            e.AllocatedEndpoint = new(e, "localhost", ports[e.Name], targetPortExpression: $$$"""{{- portForServing "{{{e.Name}}}" -}}""");
```
Comment (karolz-ms) [inline tests/Aspire.Hosting.Tests/Dapr/DaprTests.cs:156]: Is it necessary to set the `targetPortsExpression` here?

**Review by karolz-ms — COMMENTED** [review]
Looks great overall, thanks so much @paule96 for taking care of this. I have just one question regarding the use of `targetPortExpression` in the test

**Review by davidfowl — APPROVED** [review]

**Review by karolz-ms — APPROVED** [review]

File: tests/Aspire.Hosting.Tests/Dapr/DaprTests.cs (hunk 16 above)
Comment (paule96) [reply tests/Aspire.Hosting.Tests/Dapr/DaprTests.cs:156]: I'm not sure if the dapr related code would work without it. I will take a look later.

Comment (paule96) [reply tests/Aspire.Hosting.Tests/Dapr/DaprTests.cs:156]: @karolz-ms  yes it is important. If you don't do that you will get the following exception inside of `Aspire.Hosting.ApplicationModel.EndpointReference`

image

Comment (karolz-ms) [reply tests/Aspire.Hosting.Tests/Dapr/DaprTests.cs:156]: OK thanks for confirmation
